sql/src/main/java/io/crate/operation/collect/sources/TableFunctionCollectSource.java

```diff
@@ -0,0 +1,125 @@
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.collect.sources;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Iterables;
+import io.crate.analyze.symbol.Reference;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.analyze.symbol.Symbols;
+import io.crate.core.collections.Bucket;
+import io.crate.core.collections.Row;
+import io.crate.metadata.Functions;
+import io.crate.metadata.ReferenceInfo;
+import io.crate.metadata.table.TableInfo;
+import io.crate.metadata.tablefunctions.TableFunctionImplementation;
+import io.crate.operation.BaseImplementationSymbolVisitor;
+import io.crate.operation.Input;
+import io.crate.operation.InputRow;
+import io.crate.operation.collect.CrateCollector;
+import io.crate.operation.collect.InputCollectExpression;
+import io.crate.operation.collect.JobCollectContext;
+import io.crate.operation.collect.RowsCollector;
+import io.crate.operation.projectors.RowReceiver;
+import io.crate.planner.node.dql.CollectPhase;
+import io.crate.planner.node.dql.TableFunctionCollectPhase;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+@Singleton
+public class TableFunctionCollectSource implements CollectSource {
+
+    private final ClusterService clusterService;
+    private final Functions functions;
+    private final TableFunctionImplementationVisitor implementationVisitor;
+
+    @Inject
+    public TableFunctionCollectSource(ClusterService clusterService, Functions functions) {
+        this.clusterService = clusterService;
+        this.functions = functions;
+        implementationVisitor = new TableFunctionImplementationVisitor(functions);
+    }
+
+    @Override
+    public Collection<CrateCollector> getCollectors(CollectPhase collectPhase,
+                                                    RowReceiver downstream,
+                                                    JobCollectContext jobCollectContext) {
+        TableFunctionCollectPhase phase = (TableFunctionCollectPhase) collectPhase;
+        TableFunctionImplementation tableFunctionSafe = functions.getTableFunctionSafe(phase.functionName());
+        TableInfo tableInfo = tableFunctionSafe.createTableInfo(clusterService, phase.arguments());
+        Bucket resultBucket = tableFunctionSafe.execute(
+                // asInputs is safe here because there are no sub-selects yet and without sub-selects
+                // no references can be passed to a table function -> everything must be evaluable to a Input
+                Symbols.asInputs(phase.arguments()));
+
```

Comment:
hm, probably we could use a cluster level implementation visitor here instead of just casting?